### PR TITLE
Add support for diagnosticSettings

### PR DESCRIPTION
**Pull request by nishant221**
Closes #[4248](https://github.com/Azure/azure-service-operator/issues/4248)

**What this PR does / why we need it**:
Add support for a new resource `DiagnosticSetting`

**Special notes for your reviewer**:

**How does this PR make you feel**: 
Excited :)

**If applicable**:
- [X] this PR contains documentation
- [X] this PR contains tests
- [X] this PR contains YAML Samples

### CODE REVIEW

File: Taskfile.yml
Comment (theunrepentantgeek) [inline Taskfile.yml:1]: Please don't reformat our Taskfile; keep your PR focused on the new resource you're contributing.

Revert this change.

File: docs/hugo/content/reference/_index.md [hunk 1]
```diff
@@ -418,6 +418,14 @@ These resource(s) are available for use in the current release of ASO. Different
...
 
+### Next Release
+
+Development of these new resources is complete and they will be available in the next release of ASO.
+
+| Resource          | ARM Version        | CRD Version          | Supported From | Sample                                                                                                                                  |
+|-------------------|--------------------|----------------------|----------------|-----------------------------------------------------------------------------------------------------------------------------------------|
+| DiagnosticSetting | 2021-05-01-preview | v1api20210501preview | v2.10.0        | [View](https://github.com/Azure/azure-service-operator/tree/main/v2/samples/insights/v1api/v1api20210501preview_diagnosticsetting.yaml) |
```
Comment (theunrepentantgeek) [inline docs/hugo/content/reference/_index.md:427]: Unfortunately, we've just cut the v2.10 release of ASO (nearly two weeks late!). Change the `$supportedFrom` tag to `v2.11.0`.

File: v2/azure-arm.yaml [hunk 2]
```diff
@@ -2431,6 +2436,19 @@ objectModelConfiguration:
       WebTestGeolocation:
         Id:
           $armReference: false
+    2021-05-01-preview:
+      DiagnosticSetting:
+        $export: true
+        $supportedFrom: v2.10.0
```
Comment (theunrepentantgeek) [inline v2/azure-arm.yaml:2442]: As mentioned above, we just missed the v2.10 release.

```suggestion
        $supportedFrom: v2.11.0
```

File: v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			Location: tc.AzureRegion,
+			AddressSpace: &network.AddressSpace{
+				AddressPrefixes: []string{"10.0.0.0/16"},
+			},
+		},
+	}
+
+	tc.CreateResourceAndWait(vnet)
+
+	vnet_name := vnet.Status.Name
```
Comment (theunrepentantgeek) [inline v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go:45]: You don't need to wait for the resource to be created to get the name - you can use `vnet.ObjectMeta.Name` to retrieve it directly after instantiation.

File: v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+	tc := globalTestContext.ForTest(t)
+
+	rg := tc.CreateTestResourceGroupAndWait()
+	rg_name := rg.Status.Name
+
+	// vnet := VMVirtualNetwork(tc, testcommon.AsOwner(rg))
+
+	vnet := &network.VirtualNetwork{
+		ObjectMeta: tc.MakeObjectMetaWithName(tc.Namer.GenerateName("vn")),
```
Comment (theunrepentantgeek) [inline v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go:33]: There's a helper method that does exactly this.

```suggestion
		ObjectMeta: tc.MakeObjectMeta("vn"),
```

File: v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+	tc.CreateResourceAndWait(acct)
+
+	subARMID := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s", tc.AzureSubscription, *rg_name, *vnet_name)
+
+	vnet_diagnosticSettings := &insights.DiagnosticSetting{
+		ObjectMeta: tc.MakeObjectMeta("diagnosticsetting"),
+		Spec: insights.DiagnosticSetting_Spec{
+			Owner: &genruntime.ArbitraryOwnerReference{
+				ARMID: subARMID,
```
Comment (theunrepentantgeek) [inline v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go:73]: You shouldn't need to use an ARMID to reference the VNet. ArbitraryOwnerReference allows referencing an owner within the cluster, and ASO will take care of creating things in the correct order.

There's a helper for this:

```
Owner: testcommon.AsOwner(vnet)
```

File: v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+				{
+					CategoryGroup: to.Ptr("allLogs"),
+					Enabled:       to.Ptr(true),
+				},
+			},
+			StorageAccountReference: tc.MakeReferenceFromResource(acct),
+		},
+	}
+
+	tc.CreateResourceAndWait(vnet_diagnosticSettings)
```
Comment (theunrepentantgeek) [inline v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go:85]: One of the features of ASO is that it takes care of sequencing - it works out the correct order to create the resources itself. This is essential - when someone uses `kubectl` to submit a YAML file containing all the resources for their application, ASO gets all the resources at once and has to be able to do the right thing.

It's important that our tests exercise this - we've found in the past some resources where some additional work was required to make this run smoothly.

Instead of calling `tc.CreateResourceAndWait()` for each resource in turn, declare _*all*_ all of the resources and then make a single call to create them all at once:

```
tc.CreateResourcesAndWait(vnet, acct, vnet_diagnosticSettings)
```

File: v2/samples/insights/v1api/v1api20210501preview_diagnosticsetting.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+apiVersion: insights.azure.com/v1api20210501preview
```
Comment (theunrepentantgeek) [inline v2/samples/insights/v1api/v1api20210501preview_diagnosticsetting.yaml:1]: You'll need to make a recording of the sample - see [our docs](https://azure.github.io/azure-service-operator/contributing/add-a-new-code-generated-resource/#run-test-for-added-sample-and-commit-the-recording) for how to do this. 

We have this step to guarantee that our samples all work - most of us have had the experience of trying out a sample and finding out it doesn't work because of a typo, or because it's old, that sort of thing. This helps us to avoid that.

**Review by theunrepentantgeek — COMMENTED** [review]
This is really solid, good job. I made a few comments to improve your test, and you need to make a recording of your sample, but other than that it's very nearly complete.

File: Taskfile.yml
Comment (nishant221) [reply Taskfile.yml:1]: Thanks for the feedback. Reverted the change.

File: docs/hugo/content/reference/_index.md (hunk 1 above)
Comment (nishant221) [reply docs/hugo/content/reference/_index.md]: Updated to `v2.11.0`

File: v2/azure-arm.yaml (hunk 2 above)
Comment (nishant221) [reply v2/azure-arm.yaml]: Updated to `v2.11.0`

File: v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go (hunk 3 above)
Comment (nishant221) [reply v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go]: Updated with the suggested changes.

File: v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go (hunk 4 above)
Comment (nishant221) [reply v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go]: Updated with the suggested changes.

File: v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go (hunk 6 above)
Comment (nishant221) [reply v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go]: Updated with the suggested changes.

File: v2/samples/insights/v1api/v1api20210501preview_diagnosticsetting.yaml (hunk 7 above)
Comment (nishant221) [reply v2/samples/insights/v1api/v1api20210501preview_diagnosticsetting.yaml]: Added the recording of the sample.

File: v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go (hunk 5 above)
Comment (nishant221) [reply v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go]: Updated with the reference rather than ARMID.

**Comment by theunrepentantgeek** [comment]
CI is failing because you're missing an update to one file - `v2/api/insights/versions_matrix.md`.

Once that's included, we can run the full test suite and look at merging this contribution. Thanks!

Spoiler alert: I ran the full test suite locally, and it passed. 😉

**Comment by nishant221** [comment]
@theunrepentantgeek Thanks for the valuable feedbacks. Added the missing updates to the file - `v2/api/insights/versions_matrix.md` . Kindly have a re-look.
Thanks

**Review by matthchr — APPROVED** [review]

**Comment by theunrepentantgeek** [comment]
A couple of markdown files are generating with different content, looks like the generator is now picking up on the sample and is including links in the index files. I've confirmed this locally:

``` bash
$ git status
On branch feature/diagnosticsetting
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   docs/hugo/content/reference/_index.md
        modified:   docs/hugo/content/reference/insights/_index.md
```

Run `task controller:generate-types` to regenerate them. Once CI is clean, we'll merge this.

**Comment by nishant221** [comment]
Ran the above mentioned command and committed the changes  https://github.com/Azure/azure-service-operator/pull/4363/commits/0f45678862ad663ce507d52f6bb141dd01e74bde
